snapcraft/plugins/v2/cmake.py

```diff
@@ -48,27 +48,43 @@ def get_schema(cls) -> Dict[str, Any]:
                     "uniqueItems": True,
                     "items": {"type": "string"},
                     "default": [],
+                },
+                "cmake-generator": {
+                    "type": "string",
+                    "default": "Unix Makefiles",
```

Comment:
Maybe this can be an enum of supported strings?

Reply:
Actually, I don't know. CMake maintains a list of generators as constants in their codebase. Here it is:

```
* Unix Makefiles               = Generates standard UNIX makefiles.
  Green Hills MULTI            = Generates Green Hills MULTI files
                                 (experimental, work-in-progress).
  Ninja                        = Generates build.ninja files.
  Watcom WMake                 = Generates Watcom WMake makefiles.
  CodeBlocks - Ninja           = Generates CodeBlocks project files.
  CodeBlocks - Unix Makefiles  = Generates CodeBlocks project files.
  CodeLite - Ninja             = Generates CodeLite project files.
  CodeLite - Unix Makefiles    = Generates CodeLite project files.
  Sublime Text 2 - Ninja       = Generates Sublime Text 2 project files.
  Sublime Text 2 - Unix Makefiles
                               = Generates Sublime Text 2 project files.
  Kate - Ninja                 = Generates Kate project files.
  Kate - Unix Makefiles        = Generates Kate project files.
  Eclipse CDT4 - Ninja         = Generates Eclipse CDT 4.0 project files.
  Eclipse CDT4 - Unix Makefiles= Generates Eclipse CDT 4.0 project files.

```

The most useful are `Unix Makefiles` -- just plain-old make and `Ninja`.
I don't think, that anyone uses `Watcom WMake` with CMake for snap projects. And if they do, they can always just use the `nil` plugin or submit a PR here.
`Green Hills MULTI` looks like an IDE for embedded systems, and these systems certainly don't support snap, so, IMO, its native support in this plugin is not necessary, too.
`CodeBlocks`, `CodeLite`, `Sublime Text 2`, `Kate` and `Eclipse CDT4` variants are project generators for use with various IDEs, and they should not be used for CI builds in valid CMake-based projects.

So, I think, we can use an enum with a full comparison and raise an exception, if an unsupported generator is encountered.

Reply:
Is there any particular reason a snap developer would want to pick a particular generator?  My understanding was that all backends are intended to produce equivalent builds.

I understand that there are performance benefits to using Ninja over traditional Makefiles though, so is there any reason why we wouldn't want to build all cmake parts using Ninja?

Reply:
Interesting point - I would not expect there to be a problem, but I don't know it for a fact.  I would consider:

1. Projects can set their own generators - setting a non-standard default immediately puts us in potential conflict unless we try to detect it.  If there is an edge case and the project knows best, they might be justified in specifying....

2. I expect ninja would add a dependency on `ninja-build`, though very minor in the scheme of things.

In general, if it's not the build tool's default, I would avoid it being the snapcraft default without good reason.  It tends to lead to issues down the road... just my opinion :D

Reply:
> Actually, I don't know. CMake maintains a list of generators as constants in their codebase. Here it is:
> 
> ```
> * Unix Makefiles               = Generates standard UNIX makefiles.
>   Green Hills MULTI            = Generates Green Hills MULTI files
>                                  (experimental, work-in-progress).
>   Ninja                        = Generates build.ninja files.
>   Watcom WMake                 = Generates Watcom WMake makefiles.
>   CodeBlocks - Ninja           = Generates CodeBlocks project files.
>   CodeBlocks - Unix Makefiles  = Generates CodeBlocks project files.
>   CodeLite - Ninja             = Generates CodeLite project files.
>   CodeLite - Unix Makefiles    = Generates CodeLite project files.
>   Sublime Text 2 - Ninja       = Generates Sublime Text 2 project files.
>   Sublime Text 2 - Unix Makefiles
>                                = Generates Sublime Text 2 project files.
>   Kate - Ninja                 = Generates Kate project files.
>   Kate - Unix Makefiles        = Generates Kate project files.
>   Eclipse CDT4 - Ninja         = Generates Eclipse CDT 4.0 project files.
>   Eclipse CDT4 - Unix Makefiles= Generates Eclipse CDT 4.0 project files.
> ```
> 
> The most useful are `Unix Makefiles` -- just plain-old make and `Ninja`.
> I don't think, that anyone uses `Watcom WMake` with CMake for snap projects. And if they do, they can always just use the `nil` plugin or submit a PR here.
> `Green Hills MULTI` looks like an IDE for embedded systems, and these systems certainly don't support snap, so, IMO, its native support in this plugin is not necessary, too.
> `CodeBlocks`, `CodeLite`, `Sublime Text 2`, `Kate` and `Eclipse CDT4` variants are project generators for use with various IDEs, and they should not be used for CI builds in valid CMake-based projects.
> 
> So, I think, we can use an enum with a full comparison and raise an exception, if an unsupported generator is encountered.

I'd limit it to the use cases of interesting.  makefiles and ninja.